wildfire/goes/downloader.py

```diff
@@ -44,3 +47,150 @@ def download_scan(s3_bucket, s3_key, local_directory):
+        List of downloaded filepaths.
+    """
+    s3_scans = query_s3(
+        satellite=satellite, regions=regions, channels=channels, start=start, end=end
+    )
+    filepaths = []
+    for s3_scan in s3_scans:
+        filepaths.append(
+            download_scan(
+                s3_bucket=s3_scan.bucket_name,
+                s3_key=s3_scan.key,
+                local_directory=local_directory,
+            )
+        )
+    return filepaths
+
+
+def _is_good_object(key, regions, channels, start, end):
+    """Check if object should be downloaded, given parameters.
+
+    Check that the object is in the desired regions, channels, and date range.
+
+    Parameters
+    ----------
+    key : str
+        S3 key.
+    regions : list[str]
+    channels : list[int]
+    start : datetime.datetime
+        Start of date range.
+    end : datetime.datetime
+        End of date range.
+
+    Returns
+    -------
+    Boolean
+    """
+    region, channel, _, scan_started_at = utilities.parse_filepath(filepath=key)
+    return (
+        (region in regions)
+        and (channel in channels)
+        and (start <= scan_started_at <= end)
+    )
+
+
+def _num_hours_to_check(start, end):
+    """Calculate the number of hours of scand to retrieve from Amazon S3.
+
+    Parameters
+    ----------
+    start : datetime.datetime
+        Start of date range.
+    end : datetime.datetime
+        End of date range.
+
+    Returns
+    -------
+    int
+        The number of hours to retrieve from Amazon S3 rounded up to nearest hour. This is
+        primarily used as a range for tqdm's progressbar.
+    """
+    return math.ceil((end - start).total_seconds() / 3600)
+
+
+def query_s3(satellite, regions, channels, start, end):
+    """Query Amazon S3 for data files that match the input.
+
+    Parameters
+    ----------
+    satellite : str
+        Amazon S3 bucket name. Either noaa-goes16 or noaa-goes17
+    regions : list[str]
+        "F", "C", "M1", or "M2"
+    channels : list[int]
+        1 - 16
+    start : datetime.datetime
+    end : datetime.datetime
+
+    Returns
+    -------
+    list[boto3.resources.factory.s3.ObjectSummary]
+    """
+    _logger.info(
+        """Querying for s3 objects with the following properties:
+    satellite: %s
+    regions: %s
+    channels: %s
+    start date: %s
+    end date: %s""",
+        satellite,
+        regions,
+        channels,
+        start,
+        end,
+    )
+    s3 = boto3.resource("s3")
+    key_path_format = "{product_description}/{year}/{day_of_year}/{hour}/"
+    product_description_format = "ABI-L1b-Rad{region}"
+
+    scans = []
+    for region in tqdm(regions, desc="Regions"):
+        # only use the first character from region (M1 or M2 -> M)
+        product_description = product_description_format.format(region=region[0])
+
+        current = start.replace(minute=0, second=0, microsecond=0)
+        inner_pbar = tqdm(total=_num_hours_to_check(start=current, end=end), desc="Hours")
```

Comment:
This variable name seems confusing to me, why "pbar"?

Reply:
Yea! I can expand it to `progress_bar`

Reply:
done!